docs/MUSIC_COMPOSITION_ENVIRONMENT.md

```diff
@@ -0,0 +1,661 @@
+# Music Composition Environment
+
+## Vision
+
+Enable Cadenza to power an algorithmic music composition system that combines the best of live coding, offline rendering, and precise musical control. The system integrates with the [bach](https://github.com/camshaft/bach) discrete event simulator and [euphony](https://github.com/camshaft/euphony-rs) project to create a comprehensive music programming environment.
+
+Key capabilities:
+1. **Live Performance Mode**: Real-time execution responding to MIDI/UI events
+2. **Offline Rendering**: Non-realtime parallel rendering as fast as possible
+3. **DSP Precompilation**: Performance-critical audio processing compiled to native code
+4. **Interpreted Control**: High-level composition logic interpreted for flexibility
+5. **Autonomous Tasks**: Multiple concurrent tasks coordinating via discrete event simulation
+6. **Interactive REPL**: Live experimentation and parameter adjustment
+
+## Goals
+
+### Primary Goals
+
+1. **Dual Execution Modes**
+   - **Online (Live)**: Real-time event handling with predictable latency
+   - **Offline (Render)**: Fast parallel rendering without real-time constraints
+   - Switch between modes seamlessly
+   - Share composition code between both modes
+
+2. **Hybrid Compilation Strategy**
+   - Precompile DSP algorithms to native code for performance
+   - Interpret control logic for flexibility and live updates
+   - JIT compilation for hot paths when beneficial
+   - Zero-cost abstractions where possible
+
+3. **Discrete Event Simulation**
+   - Integration with bach library for event scheduling
+   - Autonomous tasks emitting commands and MIDI events
+   - Precise timing control with musical time units
+   - Task coordination and synchronization
+
+4. **Interactive Development**
+   - REPL for live coding and parameter tweaking
+   - Hot-reload of composition changes
+   - Visual feedback (waveforms, spectrograms, parameter graphs)
+   - Undo/redo for experimentation
+
+5. **Musical Units and Types**
+   - Time: beats, measures, seconds, samples
+   - Pitch: Hz, MIDI notes, note names (C4, A440)
```

Comment:
i want modes and intervals as well. so you can talk abstractly about a melody and then add more and more information to it by defining a mode that the interval operates in and then a tonic to give it an actual root. then you'd have a tuning that defines how to translate those into frequencies. but we'd want to have a bunch of predefined helpers for things that people are familiar with e.g. C4, which would expand to the appropriate type indicating how much more information is needed to turn that into an actual frequency. and this is where dimensional analysis is going to help a lot.